rules/Renaming/Rector/Name/RenameClassRector.php

```diff
@@ -130,61 +108,4 @@ public function configure(array $configuration): void
 
         $this->renamedClassesDataCollector->addOldToNewClasses($configuration);
     }
-
-    /**
-     * @param Node[] $nodes
-     * @return null|Node[]
-     */
-    public function afterTraverse(array $nodes): ?array
```

Comment:
This need to be reverted, as we rename ClassLike, from no-namespace to namepaced class, FileWithoutNamespace stmts need to be restructure as namespaced based stmts from top to down, not only single [FileWithoutNamespace]

Reply:
No, the class should never be moved in the first place. This rule should not move classes to different namespaces, that what shouldbe done in IDE/manually. Rector should only help with references.

Reply:
This change make `ClassLike::class` node types buggy, as it used by ClassRenamer service, and its stmts not pointed to correct location, eg; overlapped declare(strict_types=1)

What reference do you mean? Type hint and docblock only?

Reply:
> What reference do you mean? Type hint and docblock only?

Yes, as Rector should not be used to rename a local class. That's what developer of framework/package does. Including updating references in autoload with following work. We should only help with code changes upgrades, not replace IDE + manual work. That's why I removed previous rules that were moving classes and namespace rules.

Reply:
On ClassLike_, it can rename its class implements,eg:

```diff
-class A implements Old_
+class A implements New_
```

which valid, eg on our laminas-servicemanager-migration use case.

https://github.com/laminas/laminas-servicemanager-migration/blob/d114bcdc9361d85f1606563a0b22777c82dff653/config/rector/set/config.php#L11

That can be marked as "reference" or not?

https://github.com/rectorphp/rector-src/blob/be53faca08c2f5c3ebafb636e766788676f54d28/rules/Renaming/NodeManipulator/ClassRenamer.php#L205

partially rename probably work, but that original class name position should be kept, possibly on this case imo, and the position of stmts is essential for this case, eg, used on PostRector with rename+auto import

Reply:
Yes, you're right. This PR only remove `class X` and `namespace X` moving.

The `implements X` refrence rename still works, see: https://github.com/rectorphp/rector-src/blob/main/rules-tests/Renaming/Rector/Name/RenameClassRector/Fixture/do_not_implement_interface_twice.php.inc

<br>

If that's not the case, could you add fixture? I'll fix it :wink:  :+1:

Reply:
This part is no longer works as expected as it rename `class X` to `class Y`, on different namespace, scope is pointed to incorrect location now, and make invalid pointer due to removal of this `afterTraverse()`

https://github.com/rectorphp/rector-src/blob/be53faca08c2f5c3ebafb636e766788676f54d28/rules/Renaming/NodeManipulator/ClassRenamer.php#L207-L239

Reply:
The rename `class X` should never have happened, as it would break autoloading. We do not handle file names and location changes. 

That's scope of IDE/manual rename.

Reply:
imo, rename class like still valid, even rename namespace, with back of this `afterTraverse()`,eg:

```php
class AnotherMyOldClass
{
    /**
     * @return AnotherMyOldClass
     */
    public function createSelf(): AnotherMyOldClass
    {
        return new AnotherMyOldClass;
    }
}
```

this change to:

```diff
-class AnotherMyNewClass
+class AnotherMyOldClass
{
    /**
-     * @return AnotherMyOldClass
+     * @return \AnotherMyNewClass
     */
-    public function createSelf(): AnotherMyOldClass
+    public function createSelf(): \AnotherMyNewClass
    {
-        return new AnotherMyOldClass;
+        return new \AnotherMyNewClass;
    }
}
```

changing only reference is invalid, it needs change the class itself, but with this `afterTraverse()` need to be reverted.

Reply:
This would be the part of IDE handling. We focus on code upgrade, where class references are changed.

Reply:
Ok, then code below :

```php
$this->renameClassImplements($classLike, $oldToNewClasses, $scope);
```

should be removed, as now no longer working correctly.

Reply:
I created new PR to remove that https://github.com/rectorphp/rector-src/pull/5488